fuel/datasets/hdf5.py

```diff
@@ -495,87 +467,60 @@ def load(self):
         self._out_of_memory_open()
         handle = self._file_handle
 
-        # Load subset slices / indices
-        subsets = []
-        if len(self.which_sets) > 1:
-            indices = defaultdict(list)
-            for split in self.which_sets:
-                split_start_stop = self.get_start_stop(handle, split)
-                split_indices = self.get_indices(handle, split)
-                for source_name in self.sources:
-                    if source_name in split_indices:
-                        ind = split_indices[source_name]
-                    else:
-                        ind = list(range(*split_start_stop[source_name]))
-                    indices[source_name] = sorted(
-                        set(indices[source_name] + ind))
-        else:
-            start_stop = self.get_start_stop(handle, self.which_sets[0])
-            indices = self.get_indices(handle, self.which_sets[0])
-        num_examples = None
-        for source_name in self.sources:
-            subset = self._subset_template
-            # If subset has a step greater than 1, we convert it to a list,
-            # otherwise we won't be able to take that subset within the context
-            # of a split defined by a slice.
-            if hasattr(subset, 'step') and subset.step not in (1, None):
-                subset = list(range(len(handle[source_name])))[subset]
-            self._subset_template = subset
-            if source_name in indices:
-                source_subset = indices[source_name]
-            else:
-                source_subset = slice(*start_stop[source_name])
-            if hasattr(subset, 'step') and hasattr(source_subset, 'step'):
-                subset = slice(
-                    source_subset.start
-                    if subset.start is None else subset.start,
-                    source_subset.stop
-                    if subset.stop is None else subset.stop,
-                    subset.step)
-                subsets.append(subset)
-                subset_num_examples = subset.stop - subset.start
-            else:
-                if hasattr(source_subset, 'step'):
-                    source_subset = numpy.arange(
-                        source_subset.start, source_subset.stop)
-                subset = source_subset[subset]
-                subsets.append(subset)
-                subset_num_examples = len(subset)
-            if num_examples is None:
-                num_examples = subset_num_examples
-            if num_examples != subset_num_examples:
-                raise ValueError("sources have different lengths")
-        self.subsets = subsets
-
-        # Load data sources and source shapes
+        # Infer subsets based on `which_sets`
+        subsets = [Subset.empty_subset(len(handle[source_name]))
+                   for source_name in self.sources]
+        for split in self.which_sets:
+            start_stop = self.get_start_stop(handle, split)
```

Comment:
Given that this is the only place in `hdf5.py` where `get_start_stop` and `get_indices` are used, I would even consider replacing them with `get_subset` in this PR.